Enable csv import of partners with six fields

McEileen opened the pull request:
Resolves #5056

### Description
- Importing partners will now include more fields.
- To manually test this: 
1. log in as a bank organization admin (org_admin1@example.com)
2. navigate to `/partners` and click "Import partners" 
3. then upload a csv file with the following fields: `name,email,default_storage_location,send_reminders,quota,notes`
- I would appreciate feedback on supporting the upload of a default storage location.  I added error handling for scenarios where a user enters a default storage location that is associated with a different organization.  I tried to add error handling for scenarios where a user enters a default storage location that doesn't exist, but I couldn't get the validation to work with the non-nullable `default_storage_id` column.
- Some of the new csv files specify that they include six fields.  Thoughts on this naming?  Is it helpful to be this explicit, or is it unnecessary?

### Type of change

* New feature (non-breaking change which adds functionality)


### How Has This Been Tested?
- I added unit and request specs.

McEileen committed:
enable csv import of partners with six fields

cielf reviewed (changes requested):
Thanks for taking this on -- here's your Functional review, pass one.  Mentioning @dorner re your above request for advice.

Review notes:  
1/  I'd like to see a warning if the default storage location indicated does not exist as a storage location on the organization -- I think uploading, but with a warning would be the case of least surprise here.
2/  I left the send_reminders blank on a partner.   I expected that to give a false send_reminders,  but instead the partner failed to upload, with no explanation.
3/  I then added a few more items to the upload file,  including blank default storage partner.   I got an "ActiveModel::UnknownAttributeError at /partners/import_csv.  
unknown attribute 'default_storage_location' for Partner."   error on the blank one.    If there is no default storage location indicated,  the partner should be uploadable - we just don't set it.
4/  I like the case insensitive matching on the storage locations!

dorner commented:
@McEileen when you say you couldn't get it working... what did you try, and what problems did you see?

McEileen commented:
@dorner Good question.  My initial approach was to search for the default storage location using the name that the user entered, and to then throw an error if `StorageLocation.find_by(name: default_storage_location_name)` returned `nil`, since that would indicate that the storage location doesn't exist.  I thought it would be a good way to catch typos, like if someone types "Bluk Storage Location."  The problem is that this approach would require the user to only import partners that have default storage locations associated with them.  This contradicts the data model, which shows that `default_storage_location_id` could be `nil`.

To give a code example, if I change the `default_storage_location_belongs_to_organization` validator in the partner model to not allow a `nil` `default_storage_location_id`, like so
```
  def default_storage_location_belongs_to_organization
    location_ids = organization&.storage_locations&.pluck(:id)
    unless location_ids&.include?(default_storage_location_id)
      errors.add(:default_storage_location_id, "The default storage location is not a storage location for this partner's organization")
    end
  end
```

It causes 39 tests to fail.  :sweat:

McEileen commented:
@cielf Thank you for the thorough functional review!

1/ I'll work on this and reach out with any follow-up questions.

2/ Did the test csv file you upload have the `send_reminders` header and it was missing the field itself, as seen below?
```
name,email,default_storage_location,send_reminders,quota,notes
Partner 1,partner1@example.com,"Smithsonian Conservation Center",50,"great partner"
```

Or was it missing both the `send_reminders` header and field, as seen below?
```
name,email,default_storage_location,quota,notes
Partner 1,partner1@example.com,"Smithsonian Conservation Center",50,"great partner"
```

3/ Same question here.  When there's a blank default storage partner, does that mean the header is present and the field is missing, as seen below?
```
name,email,default_storage_location,send_reminders,quota,notes
Partner 1,partner1@example.com,true,50,"great partner"
```

Or does it mean both the default_storage_location header and field are missing, as seen below?
```
name,email,default_storage_location,send_reminders,quota,notes
Partner 1,partner1@example.com,"Smithsonian Conservation Center",true,50,"great partner"
```

cielf commented:
@McEileen -- 2/ I think it had the header, but for that row, it had no value.   I was working with numbers and exporting...  so it was showing blank rather than FALSE.   Which should have produced something like 
name,email,default_storage_location,send_reminders,quota,notes
Partner 1,partner1@example.com,"Smithsonian Conservation Center",,50,"great partner"

If that doesn't work, let me know and I'll reproduce an example.

3/ Similarly for this one
name,email,default_storage_location,send_reminders,quota,notes
Partner 1,partner1@example.com,,FALSE,50,"great partne

should do the trick

McEileen committed:
improve resilience of csv import of partners

McEileen commented:
@cielf @dorner I addressed outstanding feedback, and also implemented some changes that @awwaiid had suggested during a past Sunday office hours.  Please let me know if there are other changes you all would like to see.

Thank you for the thorough reviews!

cielf reviewed (dismissed):
I'm down to 1 quibble, functionally:   
If we have a partner with a misspelled storage location name,   can we get that to appear as a warning rather than an error?  (ie show it as black on yellow, rather than the white on red).  

 Otherwise looks very good!

McEileen committed:
enable importable concern to display flash errors or flash alerts

McEileen commented on app/models/partner.rb:
```diff
@@ -101,18 +102,21 @@ def approvable?
-      if svc.errors.present?
+      if svc.errors.present? && svc.partner.errors.blank?
+        warnings << "#{svc.partner.name}: #{svc.errors.full_messages.to_sentence}"
+      elsif svc.errors.present?
```
@cielf I wanted to highlight this logic to you.  

If a user attempts to upload a partner with an email and name that is already in use, in addition to a misspelled storage location name, both the email/name messages and the storage location message will display as errors.  If you want, I could look into displaying the email/name error messages as errors and the storage location message as a warning.  Let me know what you prefer.

cielf replied:
My initial take is that if the partner is not getting added, having everything as errors won't be a big deal.

McEileen reviewed (commented):
Hi @cielf, I made the warning change that you requested.  Let me know what you think!  (I am a little uneasy about how it required me to make changes to other models that also use the `importable` concern, however, I was able to get `#import_csv` tests passing.)

McEileen committed:
make warnings display as errors when partner import hits both errors and warnings

cielf commented:
Hey @McEileen -- don't know if this is supposed to be ready for review,  but partner 9 in the earlier attached csv should show a warning, and isn't.

cielf commented:
Hey @McEileen -- are you still working on this?

McEileen commented:
Hi, I apologize for being out of touch.  As I shared with greater detail in slack, I won't be able to contribute to Human Essentials for the next couple of months.  Anyone who is interested is willing to work on the linked issue, #5056, and is also welcome to build off the code in this PR.
Thank you for all the valuable feedback I received in the review process. :pray:

jonny5 committed:
Refactor warnings to be part of the mixin

- Fix specs associated with CSV import

jonny5 commented:
I have updated this PR to have the warnings display. I think it should be good for testing now. Here is screenshot from uploading the test csv `partners_import_test_20250602.csv` (@cielf )

<img width="731" height="311" alt="Screenshot 2025-09-17 at 11 41 38 AM" src="https://github.com/user-attachments/assets/46eb0161-8d4c-4f82-97c8-d6990250c7d0" />

jonny5 committed:
Fix error display

jonny5 committed:
Fix warning text

cielf commented:
Hrm.    I'm seeing different results.
If I run setup, then check with org_admin2@example.com,  I get 
<img width="1690" height="1215" alt="Screenshot 2025-09-19 at 2 31 09 PM" src="https://github.com/user-attachments/assets/5144b84a-d0fa-4871-b07b-717e737f40c1" />

If I then check with org_admin1@example.com, I get: 
<img width="1713" height="1109" alt="Screenshot 2025-09-19 at 2 32 04 PM" src="https://github.com/user-attachments/assets/565e0d0b-06b7-49dc-bddc-d762b2a16d1f" />

cielf commented:
Some thoughts:  I suspect that org_admin2@example.com doesn't have a default storage location set, whereas org_admin1@

cielf commented:
But/and this raises bigger questions.   (IIRC, we do not currently allow partners to send requests to multiple banks) so I'm ccing in @awwaiid.

cielf commented:
Also ccing in @ruestitch
For *this* PR, it looks to me like we are failing silently if the partner to be imported already exists on another bank.

scooter-dangle commented on app/controllers/concerns/importable.rb:
```diff
@@ -25,13 +25,17 @@ module Importable
     if params[:file].present?
       data = File.read(params[:file].path, encoding: "BOM|UTF-8")
-      csv = CSV.parse(data, headers: true)
+      csv = CSV.parse(data, headers: true, skip_blanks: true)
```
For consistence, do we actually want to add `skip_blanks: true` when we have a number of other places where we parse CSV data where we don't? It's easier to explain a rule on import to users that's true in all cases.

If we think we should allow+skip blanks on CSV import, can we leave that out here and do it in a separate PR updating every `CSV.parse` we're using?

jonny5 committed:
Merge branch 'main' into em/5056-import-partners

jonny5 committed:
Remove skip blanks for consistency

awwaiid committed:
Merge branch 'main' into em/5056-import-partners

awwaiid committed:
Merge branch 'main' into em/5056-import-partners

jonny5 committed:
Merge remote-tracking branch 'upstream/main' into em/5056-import-partners

jonny5 committed:
Add specs

jonny5 committed:
Merge remote-tracking branch 'upstream/main' into em/5056-import-partners

jonny5 committed:
Cleanup

cielf commented:
@jonny5 I am still seeing the same result that I reported on September 19.     To make sure we are on the same page,  please note carefully the ids used.

jonny5 committed:
Merge branch 'main' into em/5056-import-partners

jonny5 commented:
@cielf Thanks for taking another look. My knowledge of the project is limited so it's very possible I am missing something. Here is a screen record of me following the steps as I understand them from the above comments and what errors/warnings I am seeing:

https://github.com/user-attachments/assets/a1701e62-e01e-4f8e-9e23-8f95d8bf647b

cielf commented:
That *does* seem to address the problem.

awwaiid committed:
Merge branch 'main' into em/5056-import-partners

jonny5 commented:
@awwaiid
